### PR TITLE
BUG: Input matrix with axis None was flattened before using median.

### PR DESCRIPTION
**Pull request by sagoyal2**
Closes #4301

### CODE REVIEW

**Comment by eric-wieser** [comment]
As mentioned in the other issue, it would be better if this fix could go upstream into `np.ndarray.partition`

**Comment by eric-wieser** [comment]
I've pushed a fix to `partition` at #10658, although it doesn't fix this problem

**Review by eric-wieser — DISMISSED** [review]
Marking this as needing changes for the above reasons.

Note that `partition` now works as intended in master

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -3318,6 +3318,9 @@ def median(a, axis=None, out=None, overwrite_input=False, keepdims=False):
     >>> assert not np.all(a==b)
 
     """
+    if isinstance(a, np.matrix) and axis is None:
+        a = array(a)
+
```
Comment (eric-wieser) [inline numpy/lib/function_base.py]: I think this should move inside `_median`

**Comment by mattip** [comment]
@sagoyal2 do you want to pursue getting this in?

**Comment by seberg** [comment]
Going to close this, since Eric said the fix should probably be in `partition`.  We discussed it briefly yesterday, and the other point is that we are starting to discourage `np.matrix` and it would be nice to avoid special casing it.

That doesn't mean we can't do anything about the bug though.  The problem is that matrix makes an array 2D, that should be 1D (or 0D).  We could for example add a check for that and raise an error.

**Pull request closed without merging**